cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/io/cdap/cdap/etl/batch/mapreduce/MapReducePreparer.java

```diff
@@ -218,8 +219,21 @@ protected SubmitterPlugin createJoiner(BatchJoiner<?, ?, ?> batchJoiner, StageSp
 
   @Override
   protected SubmitterPlugin createAutoJoiner(BatchAutoJoiner batchJoiner, StageSpec stageSpec) {
-    // TODO: (CDAP-16709) implement auto-join for mapreduce
-    throw new UnsupportedOperationException("");
+    String stageName = stageSpec.getName();
+    ContextProvider<DefaultJoinerContext> contextProvider =
+      new JoinerContextProvider(pipelineRuntime, stageSpec, context.getAdmin());
+    return new SubmitterPlugin<>(stageName, context, batchJoiner, contextProvider, joinerContext -> {
+      if (joinerContext.getNumPartitions() != null) {
+        job.setNumReduceTasks(joinerContext.getNumPartitions());
+      }
+      hConf.set(ETLMapReduce.MAP_KEY_CLASS, StructuredRecord.class.getName());
```

Comment:
This might be a design question. Since in the old joiner, the join key and join type are generic. Are we removing that functionality for the new joiner?

Reply:
yeah, it will only work with StructuredRecord